### PR TITLE
[review] roundabout turns

### PR DESCRIPTION
**Pull request by MoKob**
Small roundabouts at intersections are often better understood when announced as actual turns.

If a roundabout is small enough to be viewed as a simple intersection, an instruction like `at the roundabout turn right` can be far clearer than `take the first exit at the roundabout`.

This pull request introduces this concept into the roundabout handling.

Contributes to: https://github.com/Project-OSRM/osrm-backend/issues/2263

- [x] define qualification criteria 
      - simple entry/exit (no segregated roads, mixed entry and exit)
      - small turn radius
      - at most a standard intersection (4 roads)
      - well distinct bearings
- [x] add testcases
- [x] handle all variations of entry/exit
- [x] refactor, squash and prepare for review
- [x] @daniel-j-h reviewed
- [x] requires merge of https://github.com/Project-OSRM/osrm-backend/pull/2301

- [ ] add description to wiki:

| turn-type | description |
| --- | --- |
| roundabout-turn | A `roundabout-turn` describes a turn at a small intersection that is modelled as a roundabout. The intersection itself is small and can be overseen as a whole. The `direction modifier` indicates the normal turn direction as we would expect at an intersection. A possible announcement would be: `At the roundabout turn left`. |

### CODE REVIEW

File: src/engine/api/json_factory.cpp [hunk 1]
```diff
@@ -34,9 +34,11 @@ const constexpr char *modifier_names[] = {"uturn",    "sharp right", "right", "s
 // translations of TurnTypes. Not all types are exposed to the outside world.
 // invalid types should never be returned as part of the API
 const constexpr char *turn_type_names[] = {
-    "invalid", "no turn", "invalid",     "new name",   "continue", "turn",        "merge",
-    "ramp",    "fork",    "end of road", "roundabout", "invalid",  "roundabout",  "invalid",
-    "rotary",  "invalid", "rotary",      "invalid",    "invalid",  "notification"};
+    "invalid", "new name",        "continue",        "turn",       "merge",      "ramp",
```
Comment (TheMarex) [inline src/engine/api/json_factory.cpp]: You are adding a new turn type here. That requires:
- [ ] Adding an entry to `CHANGELOG.md`
- [ ] Writing a draft for the documentation of that parameter (which will be added to the wiki)

File: features/guidance/rotary.feature [hunk 2]
```diff
@@ -194,15 +194,15 @@ Feature: Rotary
 
         And the ways
             | nodes | junction   |
-            | ab    |            |
+            | ad    |            |
```
Comment (TheMarex) [inline features/guidance/rotary.feature]: This changes looks unrelated. Can we break this out?

File: features/guidance/roundabout.feature [hunk 3]
```diff
@@ -308,15 +308,15 @@ Feature: Basic Roundabout
 
         And the ways
             | nodes | junction   |
-            | ab    |            |
+            | ad    |            |
```
Comment (TheMarex) [inline features/guidance/roundabout.feature]: Again unrelated change.

File: include/extractor/guidance/turn_instruction.hpp [hunk 4]
```diff
@@ -36,26 +38,30 @@ enum DirectionModifier
...
-    EnterRoundaboutAtExit,  // Entering a small Roundabout at a countable exit
-    EnterAndExitRoundabout, // Touching a roundabout
-    ExitRoundabout,         // Exiting a small Roundabout
-    EnterRotary,            // Enter a rotary
-    EnterRotaryAtExit,      // Enter A Rotary at a countable exit
-    EnterAndExitRotary,     // Touching a rotary
-    ExitRotary,             // Exit a rotary
-    StayOnRoundabout,       // Continue on Either a small or a large Roundabout
-    Notification            // Travel Mode Changes`
+    Invalid,                            // no valid turn instruction
```
Comment (TheMarex) [inline include/extractor/guidance/turn_instruction.hpp]: Changing the order here means we are breaking the data format. Please add a note to the changelog/commit.

File: features/guidance/rotary.feature (hunk 2 above)
Comment (MoKob) [inline features/guidance/rotary.feature]: https://github.com/Project-OSRM/osrm-backend/pull/2321

File: features/guidance/roundabout.feature (hunk 3 above)
Comment (MoKob) [inline features/guidance/roundabout.feature]: https://github.com/Project-OSRM/osrm-backend/pull/2321

File: include/extractor/guidance/turn_instruction.hpp (hunk 4 above)
Comment (MoKob) [inline include/extractor/guidance/turn_instruction.hpp]: added a list for 5.1.0

File: include/extractor/guidance/roundabout_handler.hpp [hunk 5]
```diff
@@ -61,17 +65,21 @@ class RoundaboutHandler : public IntersectionHandler
...
     // Produces instructions to enter/exit a roundabout or to stay on it.
     // Performs the distinction between roundabout and rotaries.
-    Intersection handleRoundabouts(const bool is_rotary,
+    Intersection handleRoundabouts(const RoundaboutType roundabout_type,
                                    const EdgeID via_edge,
                                    const bool on_roundabout,
                                    const bool can_exit_roundabout,
                                    Intersection intersection) const;
+
+    bool qualifiesAsRoundaboutIntersection(const std::set<NodeID> &roundabout_nodes) const;
```
Comment (daniel-j-h) [inline include/extractor/guidance/roundabout_handler.hpp]: Why std::set? I expected a std::vector or do you require the nodes to be unique? But even then.. Hmm can you clarify?

File: include/extractor/guidance/turn_instruction.hpp [hunk 6]
```diff
@@ -80,27 +86,45 @@ struct TurnInstruction
         return TurnInstruction(TurnType::NoTurn, DirectionModifier::UTurn);
     }
 
-    static TurnInstruction REMAIN_ROUNDABOUT(bool is_rotary, const DirectionModifier modifier)
+    static TurnInstruction REMAIN_ROUNDABOUT(const RoundaboutType, const DirectionModifier modifier)
```
Comment (daniel-j-h) [inline include/extractor/guidance/turn_instruction.hpp]: What I never understood is why they're in all caps.

File: src/engine/guidance/post_processing.cpp [hunk 7]
```diff
@@ -43,7 +47,9 @@ void print(const std::vector<RouteStep> &steps)
         for (const auto &intersection : step.maneuver.intersections)
             std::cout << "(" << intersection.duration << " " << intersection.distance << ")";
 
-        std::cout << "] name[" << step.name_id << "]: " << step.name << std::endl;
+        std::cout << "] name[" << step.name_id << "]: " << step.name
+                  << " Bearings: " << step.maneuver.bearing_before << " "
+                  << step.maneuver.bearing_after << std::endl;
```
Comment (daniel-j-h) [inline src/engine/guidance/post_processing.cpp]: What happened to our beautiful logger? In all seriousness, maybe we should split off the debug utilities into a guidance/debug.hpp header or similar.

File: src/engine/guidance/post_processing.cpp [hunk 8]
```diff
@@ -170,15 +198,55 @@ void closeOffRoundabout(const bool on_roundabout,
...
+                    //   225        135               315        45
+                    //        180                           0
+                    //
+                    // A turn from north to north-east offerst bearing 0 and 45 has to be translated
+                    // into a turn of 135 degrees.
+                    // The same holdes for 90 - 135 (east to south east).
+                    // For north, the transformation works by angle = 540 (360 + 180) - exit_bearing
+                    // % 360;
+                    // All other cases are handled by first rotating both bearings to an
+                    // entry_bearing of 0.
```
Comment (daniel-j-h) [inline src/engine/guidance/post_processing.cpp]: :+1: Beautiful

File: src/engine/guidance/post_processing.cpp [hunk 9]
```diff
@@ -308,22 +376,27 @@ void trimShortSegments(std::vector<RouteStep> &steps, LegGeometry &geometry)
...
     // steps to the front.
-    // In addition, we cannot reduce the travel time that is accumulated at a different location.
-    // As a direct implication, we have to keep the time of the initial/final turns (which adds a
-    // few seconds of inaccuracy at both ends. This is acceptable, however, since the turn should
+    // In addition, we cannot reduce the travel time that is accumulated at a different
+    // location.
+    // As a direct implication, we have to keep the time of the initial/final turns (which adds
+    // a
+    // few seconds of inaccuracy at both ends. This is acceptable, however, since the turn
+    // should
```
Comment (daniel-j-h) [inline src/engine/guidance/post_processing.cpp]: Can you re-wrap the lines, seems like clang-format breaks but does not join lines

File: src/extractor/guidance/roundabout_handler.cpp [hunk 10]
```diff
@@ -116,7 +126,83 @@ void RoundaboutHandler::invalidateExitAgainstDirection(const NodeID from_nid,
...
 }
 
-bool RoundaboutHandler::isRotary(const NodeID nid) const
+// If we want to see a roundabout as a turn, the exits have to be distinct enough to be seen a
+// dedicated turns. We are limiting it to four-way intersections with well distinct bearings.
+// All entry/roads and exit roads have to be simple. Not segregated roads.
+// Processing segregated roads would technically require an angle of the turn to be available
+// in postprocessing since we correct the turn-angle in turn-generaion.
+bool RoundaboutHandler::qualifiesAsRoundaboutIntersection(
+    const std::set<NodeID> &roundabout_nodes) const
```
Comment (daniel-j-h) [inline src/extractor/guidance/roundabout_handler.cpp]: see comment about set above ^

File: src/extractor/guidance/roundabout_handler.cpp [hunk 11]
```diff
@@ -116,7 +126,83 @@ void RoundaboutHandler::invalidateExitAgainstDirection(const NodeID from_nid,
...
+        {
+            // given the reverse edge and the forward edge on a roundabout, a simple entry/exit
+            // can only contain a single further road
+            if (node_based_graph.GetOutDegree(node) > 3)
+            {
+                return false;
+            }
+        }
+        return true;
+    }();
```
Comment (daniel-j-h) [inline src/extractor/guidance/roundabout_handler.cpp]: This is std::find_if / std::find_if_not with a lambda

File: src/extractor/guidance/roundabout_handler.cpp [hunk 12]
```diff
@@ -116,7 +126,83 @@ void RoundaboutHandler::invalidateExitAgainstDirection(const NodeID from_nid,
...
+    // Find all exit bearings. Only if they are well distinct (at least 60 degrees between
+    // them), we allow a roundabout turn
+
+    const auto exit_bearings = [this, &roundabout_nodes, getCoordinate]() {
+        std::vector<double> result;
+        for (const auto node : roundabout_nodes)
+        {
+            // given the reverse edge and the forward edge on a roundabout, a simple entry/exit
+            // can only contain a single further road
+            for (const EdgeID edge : node_based_graph.GetAdjacentEdgeRange(node))
```
Comment (daniel-j-h) [inline src/extractor/guidance/roundabout_handler.cpp]: auto (?)

File: src/extractor/guidance/roundabout_handler.cpp [hunk 13]
```diff
@@ -116,7 +126,83 @@ void RoundaboutHandler::invalidateExitAgainstDirection(const NodeID from_nid,
...
+                result.push_back(
+                    util::coordinate_calculation::bearing(src_coordinate, next_coordinate));
+                break;
+            }
+        }
+        std::sort(result.begin(), result.end());
+        return result;
+    }();
+
+    const bool well_distinct_bearings = [](const std::vector<double> &bearings) {
```
Comment (daniel-j-h) [inline src/extractor/guidance/roundabout_handler.cpp]: assert on size / indexes below

File: include/extractor/guidance/roundabout_handler.hpp (hunk 5 above)
Comment (MoKob) [inline include/extractor/guidance/roundabout_handler.hpp]: because its a broken out function. The location where it is handled requires a sorted structure that can be queried for whether it contains an element.

**Comment by MoKob** [comment]
superseeded by https://github.com/Project-OSRM/osrm-backend/pull/2327
